consensus/reactor.go

```diff
@@ -1,551 +1,561 @@
+				SendQueueCapacity:   100,
+				RecvBufferCapacity:  50 * 4096,
+				RecvMessageCapacity: maxMsgSize,
+			},
+		},
+		VoteChannel: {
+			MsgType: new(tmcons.Message),
+			Descriptor: &p2p.ChannelDescriptor{
+				ID:                  byte(VoteChannel),
+				Priority:            7,
+				SendQueueCapacity:   100,
+				RecvBufferCapacity:  100 * 100,
+				RecvMessageCapacity: maxMsgSize,
+			},
+		},
+		VoteSetBitsChannel: {
+			MsgType: new(tmcons.Message),
+			Descriptor: &p2p.ChannelDescriptor{
+				ID:                  byte(VoteSetBitsChannel),
+				Priority:            1,
+				SendQueueCapacity:   2,
+				RecvBufferCapacity:  1024,
+				RecvMessageCapacity: maxMsgSize,
+			},
+		},
+	}
 )
 
 const (
-	StateChannel       = byte(0x20)
-	DataChannel        = byte(0x21)
-	VoteChannel        = byte(0x22)
-	VoteSetBitsChannel = byte(0x23)
+	StateChannel       = p2p.ChannelID(0x20)
+	DataChannel        = p2p.ChannelID(0x21)
+	VoteChannel        = p2p.ChannelID(0x22)
+	VoteSetBitsChannel = p2p.ChannelID(0x23)
 
-	maxMsgSize = 1048576 // 1MB; NOTE/TODO: keep in sync with types.PartSet sizes.
+	maxMsgSize = 1048576 // 1MB; NOTE: keep in sync with types.PartSet sizes.
 
 	blocksToContributeToBecomeGoodPeer = 10000
 	votesToContributeToBecomeGoodPeer  = 10000
+
+	listenerIDConsensus = "consensus-reactor"
 )
 
-//-----------------------------------------------------------------------------
+type ReactorOption func(*Reactor)
 
 // Reactor defines a reactor for the consensus service.
 type Reactor struct {
-	p2p.BaseReactor // BaseService + p2p.Switch
+	service.BaseService
 
-	conS *State
+	conS     *State
+	eventBus *types.EventBus
+	Metrics  *Metrics
 
 	mtx      tmsync.RWMutex
+	peers    map[p2p.NodeID]*PeerState
 	waitSync bool
-	eventBus *types.EventBus
-
-	Metrics *Metrics
-}
-
-type ReactorOption func(*Reactor)
-
-// NewReactor returns a new Reactor with the given
-// consensusState.
-func NewReactor(consensusState *State, waitSync bool, options ...ReactorOption) *Reactor {
-	conR := &Reactor{
-		conS:     consensusState,
-		waitSync: waitSync,
-		Metrics:  NopMetrics(),
-	}
-	conR.BaseReactor = *p2p.NewBaseReactor("Consensus", conR)
-
-	for _, option := range options {
-		option(conR)
-	}
-
-	return conR
-}
 
-// OnStart implements BaseService by subscribing to events, which later will be
-// broadcasted to other peers and starting state if we're not in fast sync.
-func (conR *Reactor) OnStart() error {
-	conR.Logger.Info("Reactor ", "waitSync", conR.WaitSync())
+	stateCh       *p2p.Channel
+	dataCh        *p2p.Channel
+	voteCh        *p2p.Channel
+	voteSetBitsCh *p2p.Channel
+	peerUpdates   *p2p.PeerUpdates
+
+	// NOTE: We need a dedicated stateCloseCh channel for signaling closure of
+	// the StateChannel due to the fact that the StateChannel message handler
+	// performs a send on the VoteSetBitsChannel. This is an antipattern, so having
+	// this dedicated channel,stateCloseCh, is necessary in order to avoid data races.
```

Comment:
Also just so I understand correctly, we need the separate channel because we need to make sure we close state channel before the rest of the channels right?

Reply:
> What's the reason behind the state channel sending stuff also over the vote set bits channel? Is it a timing thing? Like only on certain conditions do we sent vote set bits?

Good question. To me, it seems like a code smell. i.e. you should only send and receive messages on dedicated channels and not mix-n-match. As to the why, unfortunately idk. This is an artifact of how this reactor was designed/implemented. Perhaps @ebuchman can elaborate on why we do this?